common/protob/messages-ethereum.proto

```diff
@@ -66,7 +66,16 @@ message EthereumSignTx {
     optional bytes data_initial_chunk = 7;  // The initial data chunk (<= 1024 bytes)
     optional uint32 data_length = 8;        // Length of transaction payload
     optional uint32 chain_id = 9;           // Chain Id for EIP 155
-    optional uint32 tx_type = 10;           // (only for Wanchain)
+    optional uint32 tx_type = 10;           // (Used for Wanchain and EIP2718)
+    optional bytes max_gas_fee = 12;        // <=256 bit unsigned big endian (in wei) - EIP 1559
+    optional bytes max_priority_fee = 13;   // <=256 bit unsigned big endian (in wei) - EIP 1559
```

Comment:
I don't know if I agree with this, I think it might make it more confusing to use the legacy field for this 🤔

Reply:
You're probably right.

A different thing worth considering, however. Maybe it would be nicer to create a new message type for either:
(a) EIP1559 transaction -- so the eip2718 type would not be included and it would be inferred from the message, or
(b) EIP2718 transaction -- so eip2718 type would need to be specified. The question is, how different do we expect different eip2718 types to be from each other?

The main reason for doing this would be type safety: we could set most fields to `required`, both for the legacy tx and the new tx. This would simplify the situation both for callers (it's obvious which data is needed and when), and for Trezor (easier to verify data correctness).

On Trezor side, very little additional work is needed from the current state, just some reorg and registering the message handler.
On trezorlib side, we would probably have separate calls in `trezorlib.ethereum`, and maybe the same `ethereum sign-tx` in `trezorlib.cli.ethereum`? So that "IF using ethereum and not a fork, AND the user doesn't explicitly specify one or the other, THEN we default to EIP1559"?

I am not sure if this doesn't make things difficult elsewhere though. What do you think?

Reply:
I think this might be overcomplicating things at this point. I expect, in the future, that new transaction types will use the EIP 1559 gas params but with another transaction type number and possibly a new param or two appended. 

I also think that the caller itself should decide the values that go into the transaction instead of the library trying to guess which transaction type you are trying to use.

Reply:
> I also think that the caller itself should decide the values that go into the transaction instead of the library trying to guess which transaction type you are trying to use.

for library usage, yes, absolutely. what I mean above is that the _CLI command_ should make this guess for the user, unless params are explicitly specified

Reply:
> I think this might be overcomplicating things at this point

I wouldn't consider it more complicated than the current implementation. There is very little code duplication between the legacy and EIP1559 signing, and what is there can be very cleanly extracted into a shared function (token handling; transaction data handling).

If in the future another type is specified that _only adds_ fields to current EIP1559 transaction, it's easy to extend `EthereumSignEIP1559Tx` with those fields, and possibly introduce a `tx_type` field if necessary (i'd infer it from the presence or absence of the new fields).

OTOH, if in the future another type is specified to _add and remove_ fields, we'll need to install more `if`s into `check()` and `get_total_length()` and we can't implement strong type checking for the fields that are there (which will allow us to, e.g., get completely rid of `sanitize()` function)

It currently seems to me that a separate `EthereumSignEIP1559Tx` operation is a cleaner solution for possible future extensions. I'm not strongly convinced, but I'd like to see at least some arguments against, if you have them :)

Reply:
It just seems to me that adding a `EthereumSignEIP1559Tx` would result in the API being trickier to use and split into a more advanced API from the callers perspective. If I am thinking of how we wanted to implement it in MyCrypto we wouldn't want to call another function to do EIP 1559, we would rather pass the correct arguments to a single function, that then signs legacy or EIP 1559 or whatever else.

But if you prefer adding the `EthereumSignEIP1559Tx` we can go that route.

Reply:
If that's what you would rather want, I might need a better explanation of how you imagine it would work. @matejcik

Reply:
Because how I imagine it, it would complicate the API here: 
https://github.com/trezor/connect/blob/f71a8e256bf8a51c9bf835291e4969d994d56aff/src/js/core/methods/EthereumSignTransaction.js#L19

Reply:
do you already have code in MyCrypto that can send either legacy or eip1559 transaction? (for some other signing method than Trezor)
can you point me to it?

Reply:
We use a common abstraction for most wallets which can be found here: https://github.com/MyCryptoHQ/wallets

Each integration implements the shared interface https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/wallet.ts

The idea is that they take a shared `TransactionRequest` type which may or may not be EIP 1559 (or something else) and signs them. The app itself would determine the parameters of the transaction and whether it is a EIP 1559 tx or legacy.

Examples:
Private Keys: https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/non-deterministic/private-key.ts#L11-L14

Ledger (WIP): https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/deterministic/ledger.ts#L28-L54

Trezor: https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/deterministic/trezor.ts#L27-L53

Adding two separate functions/messages for signing transactions would complicate it in our implementation at least. It doesn't make it impossible or anything, just is less nice to work with from my POV.

Reply:
This could very well be hidden in Trezor Connect, if it would make sense.

actually looking at [TransactionRequest](https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-provider/src.ts/index.ts#L18-L35) it might make sense to accept _that_ in Connect. Is this "ethers.js" an ecosystem-wide thing?

I'll talk to Connect folks too, see what they think.

Reply:
Maybe? I haven't looked to much into the Connect implementation, so I'm not a 100% sure of the ramifications of doing the two message types. It just looks to me as it might complicate things, hence my worries as described above.

`ethers.js` is currently what we (at MyCrypto) prefer, there are other alternatives, so I wouldn't say it is necessarily an ecosystem default, but it is used by many. I don't know if it makes sense for you guys to use it as the request type, that would be up to you.

Reply:
looking at ethers.js [signing code](https://github.com/ethers-io/ethers.js/blob/71b7547f10229f50d8b701611c5e6041b8ed966b/packages/transactions/src.ts/index.ts), it seems to come down to that at a certain point, _some_ piece of code must say:
```
if eip2718_type is None:
	check_legacy_fields(...)
	sign_legacy_tx(...)
elif eip2718_type == X:
	check_X_fields(...)
	sign_X_tx(...)
elif eip2718_type == Y:
	check_Y_fields(...)
	sign_Y_tx(...)
# etc.
```
I would prefer that this was not explicitly coded in Trezor firmware, as the "different message" is a good mechanism that does the above implicitly. Perhaps doing this in Connect would be OK.

I'll let you know once we discuss this with Connect folks.

Reply:
We discussed this with Connect and decided that on the firmware side, it is better for us to split into two messages. The Connect API will most likely be unified, the way you like to use it.

Please create a new message `EthereumSignTxEIP1559` with the appropriate fields, no `tx_type` (we will use type 2 as constant for now).

You will need to add an entry to `messages.proto` to register the new message type (pick a new id in the ~450 range), create `ethereum/sign_tx_eip1559.py`, and register it in `workflow_handlers.py`.

Other than that it should just be a matter of rearranging the code that already exists. Let me know if you encounter any problems.

Reply:
@matejcik Done! Let me know if there is any more small things otherwise I would assume this is good for a final review 😄